cmd/sign/main.go

```diff
@@ -0,0 +1,176 @@
+	s.info.Println("handler:", r.Method, r.URL)
+	path := strings.TrimPrefix(r.URL.String(), "/v2/")
+
+	switch {
+	case path == "":
+		// API Version check.
+		w.Header().Set("Docker-Distribution-API-Version", "registry/2.0")
+		return
+	case strings.Contains(path, "/blobs/"),
+		strings.Contains(path, "/manifests/sha256:"):
+		// Extract requested blob digest and redirect to serve it from GCS.
+		// If it doesn't exist, this will return 404.
+		parts := strings.Split(r.URL.Path, "/")
+		digest := parts[len(parts)-1]
+		serve.Blob(w, r, digest)
+	case strings.Contains(path, "/manifests/"):
+		s.serveSignManifest(w, r)
+	default:
+		serve.Error(w, serve.ErrNotFound)
+	}
+}
+
+// sign.kontain.me/ubuntu -> mirror ubuntu and serve
+func (s *server) serveSignManifest(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	path := strings.TrimPrefix(r.URL.Path, "/v2/")
+	parts := strings.Split(path, "/")
+
+	refstr := strings.Join(parts[:len(parts)-2], "/")
+	tagOrDigest := parts[len(parts)-1]
+	if strings.HasPrefix(tagOrDigest, "sha256:") {
+		refstr += "@" + tagOrDigest
+	} else {
+		refstr += ":" + tagOrDigest
+	}
+	for strings.HasPrefix(refstr, "sign.kontain.me/") {
+		refstr = strings.TrimPrefix(refstr, "sign.kontain.me/")
+	}
+
+	ref, err := name.ParseReference(refstr)
+	if err != nil {
+		s.error.Printf("ERROR (ParseReference(%q)): %v", refstr, err)
+		serve.Error(w, err)
+		return
+	}
+
+	revision := tagOrDigest
+
+	// If request is for image by digest, try to serve it from GCS.
+	if strings.HasPrefix(tagOrDigest, "sha256:") {
+		desc, err := s.storage.BlobExists(ctx, tagOrDigest)
+		if err != nil {
+			s.error.Printf("ERROR (storage.BlobExists): %s", err)
+			serve.Error(w, serve.ErrNotFound)
+			return
+		}
+		if r.Method == http.MethodHead {
+			w.Header().Set("Docker-Content-Digest", tagOrDigest)
+			w.Header().Set("Content-Type", string(desc.MediaType))
+			w.Header().Set("Content-Length", fmt.Sprintf("%d", desc.Size))
+			return
+		}
+		serve.Blob(w, r, tagOrDigest)
+		return
+	}
+
+	// If it's a HEAD request, and request was by digest, and we have that
+	// manifest mirrored by digest already, serve HEAD response from GCS.
+	// If it's a HEAD request and the other conditions aren't met, we'll
+	// handle this later by consulting the real registry.
+	if r.Method == http.MethodHead {
+		if d, ok := ref.(name.Digest); ok {
+			if desc, err := s.storage.BlobExists(ctx, d.DigestStr()); err == nil {
+				w.Header().Set("Docker-Content-Digest", d.DigestStr())
+				w.Header().Set("Content-Type", string(desc.MediaType))
+				w.Header().Set("Content-Length", fmt.Sprintf("%d", desc.Size))
+				return
+			}
+		}
+	}
+
+	ko := sign.KeyOpts{
+		FulcioURL:    fulcioclient.SigstorePublicServerURL,
+		RekorURL:     "https://rekor.sigstore.dev",
+		OIDCIssuer:   "https://oauth2.sigstore.dev/auth",
+		OIDCClientID: "sigstore",
+	}
+
+	err = sign.SignCmd(ctx, ko, options.RegistryOptions{}, nil, []string{refstr}, "", true, "", false, false, "")
```

Comment:
So, this command will generate and attempt to push a signature to the registry. This will fail, because this registry doesn't accept pushes, and is only a pull-through cache/mirror. We have a couple options:

1. have the registry accept pushes, but only for signatures.
2. don't have this code push the signature back to its own registry, but instead have it generate the signature and write it directly to storage, so that the user can fetch and verify it.

I have a preference for (2) but I'm open to (1) as an easy workaround to help validate the idea.

Reply:
> 1. have the registry accept pushes, but only for signatures.

We can [specify the registry](https://github.com/sigstore/cosign/blob/fe003150271d7622bc5b98f92813a99ccb8f2b67/README.md#specifying-registry) using `$COSIGN_REPOSITORY` env variable here, I guess.

> 2. don't have this code push the signature back to its own registry, but instead have it generate the signature and write it directly to storage, so that the user can fetch and verify it.

IIUC, I think cosign currently does not support such a `--skip-push` or `--output-signature` features. This requires additional work to do in the `SignCmd` to decide functionality, right? @dlorenc

Reply:
@dlorenc, yes, we are so excited about this effort, so, we can add this support to cosign if there is no specific reason for not supporting it 🙏 The main idea is that we can output the signature to the local disk or standard output instead of uploading it to the registry, so, with this, we will be able to upload this signature to the GCS Bucket for later use as @imjasonh mentioned above.

Reply:
> So, this command will generate and attempt to push a signature to the registry. This will fail, because this registry doesn't accept pushes, and is only a pull-through cache/mirror. We have a couple options:

@imjasonh I couldn't understand much, can you explain a bit more, thanks in advance?

Reply:
I think we're thinking of kontain.me too much like a regular registry, instead of imagining all the magical things it could do behind the scenes 🪄 .

kontain.me doesn't just support pushes and pulls like a regular registry (it doesn't support pushes at all), it's also capable of writing _other_ data to the registry in response to pulls, without the user being involved. So when someone pulls `sign.kontain.me/ubuntu`, we'll pull and serve `ubuntu`'s manifests/blobs like normal, but we'll _also_ magically create a `:sha256-abcdef.sig` manifest and signature blobs, so that when the user does `cosign verify sign.kontain.me/ubuntu`, the signature is already there.

This doesn't have to be done the normal way, where `cosign` CLI or `sign.SignCmd` generates and _pushes_ the signature, it can also be done by doing all stuff `cosign` does directly -- generate a manifest and signature blob, and write those to registry storage at the `:sha256-abcdef.sig` tag for pulling later.

`sign.SignCmd` is probably the wrong interface to use for this, because it (correctly) assumes it's a client talking to a registry that accepts pushes. In our case, _we are the registry_, and can write manifests and blobs directly to storage.

In order to do this, we'll have to peel back what `sign.SignCmd` does to understand it better and replicate it in kontain.me -- but that's the point of kontain.me, to learn more about how things work! 😄